ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/API.hs

```diff
@@ -0,0 +1,789 @@
+    -- | Get an empty ledger state at a requested point in the LedgerDB, if it
+    -- exists.
+  , getPastLedgerState     :: Point blk -> STM m (Maybe (l EmptyMK))
+    -- | Get the header state history for all ledger states in the LedgerDB.
+  , getHeaderStateHistory  ::
+         (l ~ ExtLedgerState blk)
+      => STM m (HeaderStateHistory blk)
+    -- | Acquire a 'Forker' at a requested @n@ blocks back from the tip.
+    --
+    -- You could view this as a rollback of @n@ blocks, and acquiring the
+    -- 'Forker' at that point.
+  , getForkerAtFromTip :: Word64 -> m (Either ExceededRollback (Forker m l blk))
+    -- | Acquire a 'Forker' at the requested point or at the tip. If the
+    -- requested point doesn't exist it will return a @StaticRight (Left pt)@.
+  , getForker  ::
+         forall a b.
+         StaticEither b () (Point blk)
+      -> STM m a
+#if __GLASGOW_HASKELL__ >= 902
+         -- ^ STM operation that we want to run in the same atomic block as the
+         -- acquisition of the LedgerDB
+#endif
+    -> m ( a
+         , StaticEither b
+            (Forker m l blk)
+            (Either (Point blk) (Forker m l blk))
+         )
+    -- | Get the references to blocks that have previously been applied.
+  , getPrevApplied :: STM m (Set (RealPoint blk))
+    -- | Add new references to blocks that have been applied.
+  , addPrevApplied :: [RealPoint blk] -> STM m ()
+    -- | Garbage collect references to old blocks that have been previously
+    -- applied.
+  , garbageCollect :: SlotNo -> STM m ()
+    -- | Get
+  , getResolveBlock :: STM m (ResolveBlock m blk)
+    -- | If the provided arguments indicate so (based on the DiskPolicy with
+    -- which this LedgerDB was opened), take a snapshot and delete stale ones.
+  , tryTakeSnapshot ::
+         (l ~ ExtLedgerState blk)
+      => Maybe (Time, Time)
+#if __GLASGOW_HASKELL__ >= 902
+         -- ^ If a snapshot has been taken already, the time at which it was
+         -- taken and the current time.
+#endif
+      -> Word64
+#if __GLASGOW_HASKELL__ >= 902
+         -- ^ How many blocks have been processed since the last snapshot.
+#endif
+      -> m SnapCounters
+    -- | Flush in-memory LedgerDB state to disk, if possible. This is a no-op
+    -- for implementations that do not need an explicit flush function.
+  , tryFlush :: m ()
+  }
+
+{-------------------------------------------------------------------------------
+  Exceptions
+-------------------------------------------------------------------------------}
+
+-- | Database error
+--
+-- Thrown upon incorrect use: invalid input.
+data LedgerDbError blk =
+      -- | The LedgerDB is closed.
+      --
+      -- This will be thrown when performing some operations on the LedgerDB. The
+      -- 'CallStack' of the operation on the LedgerDB is included in the error.
+      ClosedDBError PrettyCallStack
+      -- | A Forker is closed.
+    | ClosedForkerError PrettyCallStack
```

Comment:
I'm dubious these CallStacks are useful. Have you tried throwing an error on purpose? I think it will only say where it was thrown, not where the ledgerdb function was called, unless you add the constraint to the methods (and even in that case, it might only report the place where it was created?)

Reply:
I haven't -- I'm just copying what the ChainDB is doing